File: python/Arrays/day16/lConsecutive.py
# # This function is only for brute approach...
# def ls(arr, num):
#     for val in arr:
#         if val == num:
#             return True
#     return False



n = int(input("Enter the size of array: "))

arr = list(map(int, input(f"Enter {n} numbers: ").split()))


# # brute appraoch for (Longest Consecutive Sequence in an array)...
# longest = 1
# for i in range(n):
#     x = arr[i]
#     cnt = 1

#     while ls(arr, x + 1) == True:
#         x = x + 1
#         cnt += 1
    
#     longest = max(longest, cnt)

# print("Longest Consecutive Sequence length (brute force) =", longest)





# -> better approach...
# arr.sort()
# longest = 1
# currCnt = 0
# lastSmallest = float('-inf')

# for i in range(n):
#     if arr[i] - 1 == lastSmallest:
#         currCnt += 1
#         lastSmallest = arr[i]
#     elif arr[i] != lastSmallest:
#         currCnt = 1
#         lastSmallest = arr[i]
    
#     longest = max(longest, currCnt)

# print("Longest Consecutive Sequence length (better appraoch) =", longest)





# -> optimal appraoch...
nums = set(arr)
longest = 0

for num in nums:
    if num - 1 not in nums:
        x = num
        cnt = 1

        while x + 1 in nums:
            x += 1
            cnt += 1
        
        longest = max(longest, cnt)
    
print("Longest Consecutive Sequence length (optimal approach) =", longest)

File: java/Arrays/Day16/LConsecutive.java
package Day16;
import java.util.*;

public class LConsecutive {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of Array: ");
        int n = sc.nextInt();

        System.out.print("Enter " + n + " " + "numbers: ");
        int arr[] = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextInt();
        }



        // // brute approach for (Longest Consecutive Sequence in an array)...
        // int longest = 1;
        // for (int i = 0; i < n; i++) {
        //     int x = arr[i];
        //     int cnt = 1;

        //     while(ls(arr, x + 1) == true) {
        //         x = x + 1;
        //         cnt++;
        //     }

        //     longest = Math.max(longest, cnt);
        // }

        // System.out.println("Longest Consecutive Sequence length = " + longest);






        // // better approach...
        // Arrays.sort(arr);
        // int longest = 1;
        // int currCnt = 0;
        // int lastSmallest = Integer.MIN_VALUE;

        // for (int i = 0; i < n; i++) {
        //     if (arr[i] - 1 == lastSmallest) {
        //         currCnt++;
        //         lastSmallest = arr[i];
        //     } else if(arr[i] != lastSmallest) {
        //         currCnt = 1;
        //         lastSmallest = arr[i];
        //     }
        //     longest = Math.max(longest, currCnt);
        // }


        // System.out.println("Longest Consecutive Sequence length with better approach = " + longest);






        // // optimal approach...
        HashSet<Integer> st = new HashSet<>();
        
        for (int num : arr) {
            st.add(num);
        }

        int longest = 0;

        for (int num : st) {
            if (!st.contains(num - 1)) {
                int x = num;
                int cnt = 1;

                while (st.contains(x + 1)) {
                    x = x + 1;
                    cnt++;
                }

                longest = Math.max(longest, cnt);
            }
        }
        System.out.println("Longest Consecutive Sequence length with optimal approach = " + longest);


        sc.close();
    }










    // // this linear search code is for brute approach...
    // public static boolean ls(int arr[], int num) {
    //     for (int i = 0; i < arr.length; i++) {
    //         if (arr[i] == num) {
    //             return true;
    //         }
    //     }
    //     return false;
    // }
}